src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3030,10 +3030,24 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
         {
             var localFunc = node.Symbol;
             var localFunctionState = GetOrCreateLocalFuncUsages(localFunc);
-            // The starting state is the top state, but with captured
-            // variables set according to Joining the state at all the
-            // local function use sites
+
+            // The starting state (`state`) is the top state (maybe null) by default,
+            // but with captured variables in the bottom state (not null).
             var state = TopState();
+            Normalize(ref state);
+            state.ForEach(
+                (slot, variables) =>
+                {
+                    var symbol = variables[slot].Symbol;
+                    if (Symbol.IsCaptured(symbol, localFunc))
+                    {
+                        SetState(ref state, slot, NullableFlowState.NotNull);
```

Comment:
Is this the right place to make the change? I thought there was somewhere in here where we are affirmatively setting captured variable states based on their declared type when we are visiting a local function declaration before any call sites.

Reply:
Currently the default state for local function usages (`localFunctionState.StartingState`) is created here:

https://github.com/dotnet/roslyn/blob/ffa547714f801482ac51ec98a6fbed72ec490825/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L12094-L12099

If we visit a usage, that state is updated accordingly, otherwise it's left in this default state. It doesn't seem to be setting specifically captured variables, but maybe that's what you meant?

Then when visiting the declaration, the initial state (called `state`) is the top state ("maybe null") for all variables except those captured which are updated according to the usage state (`localFunctionState.StartingState`).

I guess we could instead change the `localFunctionState.StartingState` to be initially the bottom state ("not null") hence when captured variables of `state` are joined with it, it should behave equivalently. I will try that.